include/eve/detail/function/simd/riscv/bit_cast.hpp

```diff
@@ -0,0 +1,586 @@
+  {
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_u64m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_u64m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_u64m8(a);
+  }
+  if constexpr( match(c, category::float32) )
+  {
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_f32m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_f32m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_f32m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_f32m8(a);
+  }
+  if constexpr( match(c, category::int32) )
+  {
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_i32m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_i32m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_i32m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_i32m8(a);
+  }
+  if constexpr( match(c, category::uint32) )
+  {
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_u32m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_u32m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_u32m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_u32m8(a);
+  }
+  if constexpr( match(c, category::int16) )
+  {
+    if constexpr( out_lmul == -2 ) return __riscv_vlmul_ext_i16mf2(a);
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_i16m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_i16m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_i16m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_i16m8(a);
+  }
+  if constexpr( match(c, category::uint16) )
+  {
+    if constexpr( out_lmul == -2 ) return __riscv_vlmul_ext_u16mf2(a);
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_u16m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_u16m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_u16m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_u16m8(a);
+  }
+  if constexpr( match(c, category::int8) )
+  {
+    if constexpr( out_lmul == -4 ) return __riscv_vlmul_ext_i8mf4(a);
+    if constexpr( out_lmul == -2 ) return __riscv_vlmul_ext_i8mf2(a);
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_i8m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_i8m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_i8m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_i8m8(a);
+  }
+  if constexpr( match(c, category::uint8) )
+  {
+    if constexpr( out_lmul == -4 ) return __riscv_vlmul_ext_u8mf4(a);
+    if constexpr( out_lmul == -2 ) return __riscv_vlmul_ext_u8mf2(a);
+    if constexpr( out_lmul == 1 ) return __riscv_vlmul_ext_u8m1(a);
+    if constexpr( out_lmul == 2 ) return __riscv_vlmul_ext_u8m2(a);
+    if constexpr( out_lmul == 4 ) return __riscv_vlmul_ext_u8m4(a);
+    if constexpr( out_lmul == 8 ) return __riscv_vlmul_ext_u8m8(a);
+  }
+}
```

Comment:
On a surface level, I don't think that extract should be usued in bit_cast

Reply:
This is an "extend" operation, I use it when underlying type needs to be widened to support expected type. For example, for "char" type we support using 1/8 part of real vector register (called `mf8`), but when you cast it to "int", you can at least use 1/2 part of register(`mf2`).

This is documented here: `Chapter 7. Type system` of https://github.com/riscv-non-isa/rvv-intrinsic-doc

Reply:
if they are treated as different sizes, bitcasting them wouldn't be correct.

You can't bitcast xmm into ymm in our system, for example.

If every cardinal is fundamental, why are you trying to do this?

Reply:
> , for "char" type we support using 1/8 part of real vector register (called mf8), but when you cast it to "int", you can at least use 1/2 part of register(mf2).

I'm not sure I understand what's happening here. 

4 bytes is 4 bytes. Both would be VLEN / 4 bytes. Why not?

Reply:
As rvv intrinsics for RISC-V usually do not know about vlen, they should fit for any available vlen. Because of this fact, some types that are correspond for fractional usage for real vector register  are unavailable.

For example, you can not fit `int` with usage 1/4 of vector register of vlen=64. So in rvv intrinsics interface there are no types that could be used to represent usage 1/4 of vector register with the element type of `int`.

However, for `char` you can fit at least one element with usage 1/8 of vector register even if vlen is 64. So type that corresponds to usage of 1/8 of vector register with element type of char is available.

In current review I try to use RISC-V vector registers in the most effective way, so I use the smallest possible types for each `wide`, that can result into different sizes for underlying types, even if `sizeof(T1)*N1==sizeof(T2)*N2` because of described above.